executor/src/witgen/jit/processor.rs

```diff
@@ -303,6 +293,75 @@ impl<'a, T: FieldElement> Processor<'a, T> {
         Ok(())
     }
 
+    /// Checks if we can finish witgen derivation, i.e. all requested variables are known,
+    /// all machine calls are complete and all polynomial identities are solved.
+    /// This function tries to guess some values for unknown variables if it does
+    /// not create a conflict.
+    /// If it is not able to finish, returns the list of missing requested variables.
+    fn try_to_finish<FixedEval: FixedEvaluator<T>>(
+        &self,
+        can_process: impl CanProcessCall<T>,
+        witgen: &mut WitgenInference<'a, T, FixedEval>,
+        identity_queue: IdentityQueue<'a, '_, T>,
+    ) -> Result<(), Vec<Variable>> {
+        // Check that we could derive all requested variables.
+        let missing_variables = self
+            .requested_known_vars
+            .iter()
+            .filter(|var| !witgen.is_known(var))
+            // Sort to get deterministic code.
+            .sorted()
+            .cloned()
+            .collect_vec();
+
+        let incomplete_machine_calls = self.incomplete_machine_calls(witgen);
+
+        // TODO we could first try to guess unknown variables and then re-check
+        // if all missing variables are known.
+
+        if !missing_variables.is_empty()
+            || !self.try_fix_simple_sends(
+                &incomplete_machine_calls,
+                can_process.clone(),
+                witgen,
+                identity_queue.clone(),
+            )
+        {
+            return Err(missing_variables);
+        }
+
+        // Now there are only missing identities left. The hope is that the identities
+        // are only about essentially unconstrained variables.
+
+        // Collect the relevant (e.g. not multiplied by a known zero value) unknown variables
+        // that are inside the block.
+        let unknown_variables = self
+            .unsolved_polynomial_identities_in_block(witgen)
+            .flat_map(|(id, row_offset)| {
+                let Identity::Polynomial(PolynomialIdentity { expression, .. }) = id else {
+                    unreachable!()
+                };
+                unknown_relevant_variables(expression, witgen, row_offset)
+                    .into_iter()
+                    .filter(|var| match var {
+                        Variable::WitnessCell(cell) | Variable::IntermediateCell(cell) => {
+                            cell.row_offset >= 0 && cell.row_offset < self.block_size as i32
```

Comment:
what is the reason for this filter here?

Reply:
We don't want to just randomly set variables outside of the block because it would create a mess. Usually, those are not constrained, because they get their values from the other block.

Reply:
I think this deserves a comment in the code :)

Unrelated to this PR, but I think we should be aware of the complexity that comes from irregular block shapes and maybe drop support. The solver should not need to be aware of internals of the variables (e.g. the row offset). In the future, this could be PlonK cells, or GKR input vectors or who knows what...

Reply:
There is a comment above "inside the block", but i'll also add one here :)

We need to know about the row offset, because otherwise we don't even know how many variables there are, and we need to set all of the rows of a column, even if the column doesn't occur in any identity.